Staging Sites: staging sync UI

kozer opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
thelinked issue.
-->

Related to https://github.com/Automattic/dotcom-forge/issues/3956

## Proposed Changes

This pr adds a new UI to support staging sync.

## Testing Instructions

### Test 1: Staging side
1. Create a WoA site.
2. Add a staging site.
3. While on **the staging side**, ensure that you see the following UI:
![2023-10-05T14:05:20,185922676+03:00](https://github.com/Automattic/wp-calypso/assets/497103/99d3c848-4d80-4cb8-8bae-bc6fe946a7a0)
4. Select **Push to production**. Ensure that if no options are selected, the **Synchronize** button cannot be pressed. Check one ( or more ) item and press the **Synchronize** button. Ensure that you see the following:
![2023-10-03T15:43:19,808194242+03:00](https://github.com/Automattic/wp-calypso/assets/497103/bbcff18b-6f05-444a-b013-83ad603922dc)
5. Ensure that if the **production site slug** is not typed, the confirmation button will not be enabled.
6. Type your site slug, and press synchronize.
7. Ensure, using your dev console that a payload named `options` is passed as a parameter.
8. Choose **Pull from production**. Ensure that you see the following UI:
![2023-10-05T14:07:28,405097884+03:00](https://github.com/Automattic/wp-calypso/assets/497103/0aece32e-8f48-41ef-810f-6a5b1a280ffc)
9. Press the **Synchronize** button. Ensure that you see the following:
![2023-10-03T15:54:24,935317368+03:00](https://github.com/Automattic/wp-calypso/assets/497103/7b262b9c-6ed0-4057-925f-f1f5c77f21e3)
10. Press `Synchronize` and ensure that no options are passed as a payload.
11. Ensure that you see the following screen:
![2023-10-05T14:09:05,482668175+03:00](https://github.com/Automattic/wp-calypso/assets/497103/ab4b46b4-82a2-43e8-a713-798fcd1fa3a1)


### Test 1: Production side
1. Create a WoA site.
2. Add a staging site.
3. While on **the production side**, ensure that you see the following UI:
![2023-10-05T14:16:18,408199774+03:00](https://github.com/Automattic/wp-calypso/assets/497103/cfb0ffd6-afa6-47e7-b285-8f09fd74f252)
4. Select **Pull from staging**. Ensure that if no options are selected, the **Synchronize** button cannot be pressed. Check one ( or more ) item and press the **Synchronize** button. Ensure that you see the following:
![2023-10-03T15:43:19,808194242+03:00](https://github.com/Automattic/wp-calypso/assets/497103/bbcff18b-6f05-444a-b013-83ad603922dc)
5. Ensure that if the **production site slug** is not typed, the confirmation button will not be enabled.
6. Type your site slug, and press synchronize.
7. Ensure, using your dev console that a payload named `options` is passed as a parameter.
8. Choose **push to staging**. Ensure that you see the following UI:
![2023-10-05T14:18:17,425500451+03:00](https://github.com/Automattic/wp-calypso/assets/497103/1fff84bc-a018-4e2f-a2c8-232605b6621a)
9. Press the **Synchronize** button. Ensure that you see the following:
10. Press `Synchronize` and ensure that no options are passed as a payload. 
11. Ensure that you see the following screen:
![2023-10-05T14:19:09,816662979+03:00](https://github.com/Automattic/wp-calypso/assets/497103/80e5b80d-dc3b-4b0d-a4aa-156b9c642940)



<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

kozer committed:
Staging Sites: Add staging to production sync support

kozer committed:
Initial commit

kozer committed:
Merge branch 'trunk' into add/staging_sync_ui

kozer committed:
Update: Add new draft UI

kozer committed:
Update: Cleanup code

kozer committed:
Update: fix render bug

kozer committed:
Update: Mobile style changes

kozer committed:
Update: Fix key issue

kozer committed:
Update: Revert accidental changes

kozer committed:
Merge branch 'trunk' into add/staging_sync_ui

kozer committed:
Update: Remove function body brackets

kozer committed:
Update: Add CheckboxTree tests

kozer committed:
Update: Staging Sync UI to match new designs

kozer committed:
Update: Change modal contents to match the new designs

kozer committed:
Update: Change modal to support extra content. Add production sync content

kozer committed:
Update: Refactor modals

kozer committed:
Update: Simplify logic

kozer committed:
Update: Fix sync cards

kozer committed:
Update: Remove CheckboxTree component, as its not needed

danielbachhuber commented:
@javierarce @wojtekn @Automattic/yolo I think this "Choose synchronization direction" UI is confusing and will be problematic:

<img width="542" alt="CleanShot 2023-10-03 at 15 30 01@2x" src="https://github.com/Automattic/wp-calypso/assets/36432/f8b2a5fc-acbe-477b-9321-e3b56f114084">

What if we standardize on "push"? Production would always be "Push to staging", and staging would always be "Push to production".

danielbachhuber commented on client/my-sites/hosting/staging-site-card/card-content/staging-sync-card.tsx:
```diff
@@ -0,0 +1,390 @@
+	},
+	{
+		name: 'contents',
+		label: translate( 'wp-content directory' ),
```
```suggestion
		label: translate( 'wp-content Directory' ),
```

danielbachhuber commented on client/my-sites/hosting/staging-site-card/card-content/staging-sync-card.tsx:
```diff
@@ -0,0 +1,390 @@
+	{
+		name: 'contents',
+		label: translate( 'wp-content directory' ),
+		subTitle: translate( 'excluding themes, plugins, and uploads' ),
```
```suggestion
		subTitle: translate( 'Everything in the wp-content directory, excluding themes, plugins, and uploads.' ),
```

wojtekn replied:
@kozer please apply the proposed change here too.

danielbachhuber commented on client/my-sites/hosting/staging-site-card/card-content/staging-sync-card.tsx:
```diff
@@ -0,0 +1,390 @@
+	},
+	{
+		name: 'roots',
+		label: translate( 'WordPress roots' ),
```
```suggestion
		label: translate( 'Web Root' ),
```

danielbachhuber commented on client/my-sites/hosting/staging-site-card/card-content/staging-sync-card.tsx:
```diff
@@ -0,0 +1,390 @@
+	{
+		name: 'roots',
+		label: translate( 'WordPress roots' ),
+		subTitle: translate( 'includes wp-config php and any non WordPress files' ),
```
@wojtekn Does the `wp-config.php` file actually get synced over? Doesn't this break the site?

wojtekn replied:
I've just tested it and it looks like it's not. I checked restore log and found this entry:

```
Skipping wp-config.php because the restore source is differente: f5:/wp-config.php
```

It would be that it doesn't synchronize `wp-config.php` if we restore from one site to another.

kozer replied:
What about this one @wojtekn ?
Maybe change that to `includes any non WordPress files`?

kozer replied:
UPDATE: Just saw your proposal below. Updated.

danielbachhuber commented on client/my-sites/hosting/staging-site-card/card-content/staging-sync-card.tsx:
```diff
@@ -0,0 +1,390 @@
+		name: 'sqls',
+		label: 'Site Database (SQL)',
+		subTitle: translate(
+			'Overwrite the database, posts, pages, products, and orders with staging data.'
```
```suggestion
			'Overwrite the database, including any posts, pages, products, or orders.'
```

danielbachhuber commented on client/my-sites/hosting/staging-site-card/card-content/staging-sync-card.tsx:
```diff
@@ -0,0 +1,390 @@
+								{ selectedItems.map( ( item ) => {
+									return <li key={ item }>{ item }</li>;
+								} ) }
+							</ConfirmationModalList>
```
Should we re-iterate the backup message here? "We'll automatically back up your site..."

wojtekn replied:
Reminding users and making them feel safe about it makes sense. On the other side, the message would be visible at the same time in both the popup and on the page inline. What do you think?

danielbachhuber replied:
I don't have a strong opinion about it. @javierarce ?

javierarce replied:
I prefer only to show it once. Right now, it's pretty noticeable next to the main CTA. The reason for not adding it in the modal is that it's preferable for the user to focus only on the task at hand and have fewer things to pay attention to.

danielbachhuber commented on client/my-sites/hosting/staging-site-card/card-content/staging-sync-card.tsx:
```diff
@@ -0,0 +1,390 @@
+								<a
+									href={ `/activity-log/${ siteSlug }` }
+									target="_blank"
+									rel="noopener noreferrer"
```
Hm, I think we should remove `target` and `rel` if it's an internal link.

kozer replied:
Daniel, I thought about that as well. I find it easier ( but this could be just me ) to open the page to a new tab, restore a backup, and move back to the other tab to initiate syncing. ( instead of navigating through the menu ). 
However, I investigated it a bit, and you are right. For internal links, we don't open pages in new tab. 
Thanks for your comment. This is fixed.

kozer committed:
Update: Fix type issues and message wording

kozer committed:
Update: Dont open activity log to a new page

wojtekn commented:
> What if we standardize on "push"? Production would always be "Push to staging", and staging would always be "Push to production".

Agreed. "Pull to" sounds weird, I think we could use "Pull from" or "Push to". What about using target environment on each context's button, so naturally, we would go with what we have on wpcalypso now:

On production:
- push to staging
- pull from staging

On staging:
- push to production
- pull from production

>  I think this "Choose synchronization direction" UI is confusing and will be problematic:

What if we switched to radio buttons for those two? Standard buttons, as in the current prototype, are problematic as one of the buttons should show synchronization types UI, and another should hide them.

javierarce commented:
The "to" is probably a misspelling we didn't notice until now, sorry about that.

> What if we switched to radio buttons for those two? Standard buttons, as in the current prototype, are problematic as one of the buttons should show synchronization types UI, and another should hide them.

It's totally ok to use radio buttons here. It makes the options immediately visible. But does this address what @danielbachhuber was asking? 

From this comment "What if we standardize on "push"? Production would always be "Push to staging", and staging would always be "Push to production" I thought he meant only to have one option (push), that would require switching to the other environment to perform. Maybe I misread it.

javierarce commented:
> Also, it would be nice if the staging version of the card more visually distinct from production, so you could more easily tell one from the other.

I'll work on a proposal for this too.

javierarce commented:
Here's a proposal addressing the things we've mentioned:

- Radio buttons
- Simplified directions
- A more distinct staging card

![image](https://github.com/Automattic/wp-calypso/assets/4933/7b8679e9-d643-4bea-8155-0b3fa1156dcb)

danielbachhuber commented:
> From this comment "What if we standardize on "push"? Production would always be "Push to staging", and staging would always be "Push to production" I thought he meant only to have one option (push), that would require switching to the other environment to perform.

Correct, this is my suggestion. I find the "Choose synchronization direction" setting to be really confusing.

> A more distinct staging card

Oooh, I like that!

javierarce commented:
> Correct, this is my suggestion. I find the "Choose synchronization direction" setting to be really confusing.

I'm glad I got that right :) That said, I think @wojtekn's suggestions help clarify the UI and make synchronizing back and forth easier without having to go elsewhere, so I'd give this design a try. 

> Oooh, I like that!

Fantastic, thanks. I'll tweak the design a little bit since it was a quick sketch.

danielbachhuber commented:
> That said, I think @wojtekn's suggestions help clarify the UI and make synchronizing back and forth easier without having to go elsewhere, so I'd give this design a try.

I think I'd like to argue in favor of:
1. Removing the option entirely.
2. Make the behavior "Pull from staging" and "Pull from production".

With this approach, the operation always _modifies the site you're currently on_. This makes it consistent with changing settings, publishing posts, etc.

javierarce commented:
I still think giving users the option to pick the direction is helpful and not very risky. 
But you have the final say, so let's do that then. I'll update the designs.

danielbachhuber commented:
> I still think giving users the option to pick the direction is helpful and not very risky.
> But you have the final say, so let's do that then. I'll update the designs.

Mine was just the argument in favor 😊 Not an overruling decision.

Thanks for providing the designs, though. I'll let @wojtekn make the final decision.

> Edit: here's what this could look like. I think it's more confusing to be honest.

Yeah, those mockups could be improved a bit more. The direction isn't mentioned until the button. One idea would be to include a graphic that depicts the direction of the sync (which may be useful even if we do the radio buttons).

danielbachhuber commented:
@javierarce @wojtekn Maybe we should get some HE feedback on the exact wording?

kozer committed:
Add: Sync progress component, and reducer updates

kozer committed:
Fix: fix loader wording issue and navigate bug

kozer committed:
Update: Change selectbox to radio buttons

wojtekn commented:
@javierarce, thanks for providing the design for both versions. I think that letting the user choose any direction on each environment using radio buttons is the clearest approach.

> With this approach, the operation always modifies the site you're currently on. This makes it consistent with changing settings, publishing posts, etc.

@danielbachhuber I see that, however, if we keep the action name related to the current context, it should be clear for the user what we are updating. E.g. we have:

- "Copy site" which reads "Copy [this] site"
- "Launch site" which reads "Launch [this] site"
- "Push to staging" which reads "Push [this] to staging"
- "Pull from staging" which reads "Pull [here] from staging"
- "Push to production" which reads "Push [this] to production"
- "Pull from production" which reads "Pull [here] from production"

> As an extra enhancement, perhaps the CTA could change based on the selected action. For example, if the user selects "Push to staging," the button could display "Synchronize staging," and if they choose "Pull from staging," the button could read "Synchronize production."

I think would be better to avoid changing button labels based on the selected option. Also, "Synchronize staging" doesn't clearly say if we will overwrite staging ("Synchronize to staging") or production ("Synchronize from staging"). 

@javierarce  @danielbachhuber  I'm wondering about one possible change - what do you think about flipping options order in both staging and production and making the "production -> staging" default selected option in both contexts? It would de-clutter the interface as users wouldn't see all possible options immediately on page load, and we would make the safer option the default one?

wojtekn commented on client/my-sites/hosting/staging-site-card/card-content/staging-sync-card.tsx:
```diff
@@ -0,0 +1,465 @@
+		name: 'roots',
+		label: translate( 'Web Root' ),
+		subTitle: translate(
+			'Everything in the wp-content directory, includes wp-config php and any non WordPress files'
```
```suggestion
			'Everything in the WordPress root directory, including any non-WordPress files.'
```

wojtekn commented on client/my-sites/hosting/staging-site-card/card-content/staging-sync-card.tsx:
```diff
@@ -0,0 +1,465 @@
+	},
+	{
+		name: 'uploads',
+		label: translate( 'Media Uploads' ),
```
It seems that media uploads require database synchronization, too. In the other case, by default, they won't appear on the destination site.

Would it make sense to clarify this here? From [docs](https://wordpress.com/support/restore/):

> Contents of the media library. You must also select ‘Site Database‘ for synchronized media uploads to appear.

kozer replied:
What if we keep the same wording as in Jetpack UI? Here is what will look like with that change in place.
![2023-10-05T14:35:19,403459067+03:00](https://github.com/Automattic/wp-calypso/assets/497103/1b4ef943-39de-4208-babf-e07909e47474)

kozer replied:
After discussions with Wojtek, we will handle this on a later iteration. Resolved for now

wojtekn commented on client/my-sites/hosting/staging-site-card/card-content/staging-sync-card.tsx:
```diff
@@ -0,0 +1,465 @@
+		name: 'sqls',
+		label: 'Site Database (SQL)',
+		subTitle: translate(
+			'Overwrite the database, including any posts, pages, products, or orders with staging data.'
```
As we use this text in both contexts, let's unify it:

```suggestion
			'Overwrite the database, including any posts, pages, products, or orders.'
```

wojtekn reviewed (commented):
@kozer looking good! I've added a few comments to improve the wording.

Let's also implement visuals:
- road work style top bar on Staging site context
- outline around staging site box on Production site context

kozer committed:
Add: Add tests for reducer

kozer committed:
Fix: radio buttons style

wojtekn commented on client/my-sites/hosting/staging-site-card/use-staging-sync.ts:
```diff
@@ -43,14 +46,21 @@ export const usePushToStagingMutation = (
+					path: `/sites/${ productionSiteId }/staging-site/pull-from-staging/${ stagingSiteId }`,
+					apiNamespace: 'wpcom/v2',
+				},
+				{ sync_options: options }
```
I cleaned this name from `sync_options` to just `options` on the REST endpoint side, please adjust it here too.

javierarce commented:
> @javierarce @danielbachhuber I'm wondering about one possible change - what do you think about flipping options order in both staging and production and making the "production -> staging" default selected option in both contexts? It would de-clutter the interface as users wouldn't see all possible options immediately on page load, and we would make the safer option the default one?

Just to clarify, you are proposing to have dropdowns with two options, and 'Push to production' is always the selected one, no matter the environment the user is seeing. That sounds good to me.

danielbachhuber commented:
@wojtekn @javierarce I don't think we should have the option selected by default. The user should have to select each option, as an extra layer of safety for gathering intent.

wojtekn commented:
> Just to clarify, you are proposing to have dropdowns with two options, and 'Push to production' is always the selected one, no matter the environment the user is seeing. That sounds good to me.

I meant to change the order of radio button options to always show production to staging direction option ("Push to staging" on production and "Pull from production" on staging site) first and make it selected by default. As this direction doesn't have any options to choose from, the default state of the interface would be simpler. Note that at this point it's just an idea I'm exploring and wanted to hear your opinion on this one.

wojtekn commented:
> @wojtekn @javierarce I don't think we should have the option selected by default. The user should have to select each option, as an extra layer of safety for gathering intent.

So it could solve the problem I'm trying to solve even without reordering options:
- by default, none of the radio buttons is selected, so additional synchronization type options are hidden
- user selects "Push to production", detailed part of the form appears
- user selects "Pull from production", detailed part of the form disappears

kozer committed:
Fix: update option name

javierarce commented:
Perfect, let's do that. I'll update the designs.

kozer committed:
Update: Fix test failure and wording

kozer committed:
Update: Dont pre-select any options by default

kozer commented:
> * ailed part of the form appears
> * user selects "Pull from production", detailed part of the form disappears

Updated the pr to include those changes.

javierarce commented:
Here are the designs for both sides:

#### Production side 
6H9TZJNUl2sqTKO7iu49vC-fi-1082_1373

![image](https://github.com/Automattic/wp-calypso/assets/4933/2363e23e-eea0-4d0c-a579-c6c0f8266d26)

#### Staging side
6H9TZJNUl2sqTKO7iu49vC-fi-1336_3409

![image](https://github.com/Automattic/wp-calypso/assets/4933/1834e918-09eb-4e8e-b439-456532d3d388)

kozer committed:
Fix: lint issues

kozer committed:
Update: Fix label wording

sejas commented:
@kozer , Great work!

I have a non-blocking suggestion to improve the UX.
I feel the `Web Root` toggle option conflicts with the rest of the options. Is the web root recursive, and does it include themes and plugins?
If so, It would be great to `auto toggle on` them or disable/hide the other options when `Web Root` is on.
The same logic should be applied for `wp-content` and `themes`/`plugins`.

https://github.com/Automattic/wp-calypso/assets/779993/33f3ce90-e210-47ad-b63d-9ab24622bb80

The confirming modal also doesn't fix the confusion problem:
<img width="872" alt="Screenshot 2023-10-05 at 16 38 09" src="https://github.com/Automattic/wp-calypso/assets/779993/0b0c13a2-f45e-4993-a5e2-5e97e68d9119">

kozer committed:
Add: Error handling support

kozer committed:
Fix: Failing tests

kozer committed:
Fix: reducer warning

sejas reviewed (commented):
@kozer , I've found a small bug.
The sentence `We are updating your production site. We’ll email you once it is ready.` changes to `We are updating your staging site. We’ll email you once it is ready.` after reloading the page. The word `production` has changed to `staging`.

**Testing instructions**

- Go to `/hosting-config/$siteUrl` In your atomic site with an staging site
- Enable any toggle and click on `Synchronize`
- Observe the progress bar and the message below: `We are updating your production site. We’ll email you once it is ready.`
- Refresh the page using the browser.
- Observe the wrong message now mentions `staging` instead of `production`.

**Screencast**

https://github.com/Automattic/wp-calypso/assets/779993/5c822866-eb1a-4de6-a9a4-d2df64af6ea0

wojtekn commented:
@sejas thanks for giving it a try!

> I feel the `Web Root` toggle option conflicts with the rest of the options. Is the web root recursive, and does it include themes and plugins?
> If so, It would be great to `auto toggle on` them or disable/hide the other options when `Web Root` is on.
> The same logic should be applied for `wp-content` and `themes`/`plugins`.

Those options don't have a common, recursive, scope:
- "roots" synchronizes files and directories from the web root, but not `wp-content/` or other core WP directories
- "wp-content" synchronizes files and directories from `wp-content/` but not `plugins/`, `themes/` or `uploads/`

To follow the previous label style, we could say:

> Everything in the WordPress root, excluding wp-content and core WordPress files.

What do you think?

> The confirming modal also doesn't fix the confusion problem:

Agreed, @kozer let's use the same option names and descriptions in the confirmation popup.

kozer committed:
Update: remove leftovers

kozer commented:
> Agreed, @kozer let's use the same option names and descriptions in the confirmation popup.
So instead of roots, etc. to use `Web Root` with the subtitle?

wojtekn commented:
@kozer correct, or at least "Web Root" and we can review if we should repeat subtitle later.

kozer commented:
Great. So I ll work on this and to the bug Antonio mentioned, do a last review and ping you

kozer committed:
Update: Change selected option on modal to use the labels instead

kozer committed:
Fix: sync loading wording

- Keep the type of the site we are using in the redux state
- Fix wording on selectors

kozer committed:
Merge branch 'trunk' into add/staging_sync_ui

kozer committed:
Fix: lint issues

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/9255494

Thank you @kozer for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/9255494).
